subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ValidateStep.java

```diff
@@ -68,7 +83,55 @@ public R execute(UnitOfWork work, C context) {
                     .collect(Collectors.toList())
             );
         }
-        return delegate.execute(work, context);
+
+        if (!warnings.isEmpty()) {
+            LOGGER.info("Invalidating VFS because of invalid work {}", work.getDisplayName());
+            virtualFileSystem.invalidateAll();
```

Comment:
Instead of invalidating all of the VFS, shall we maybe only invalidate the VFS associated with the root project or the project root for tasks?

Moreover, can you add an integration test for invalidating the VFS? Probably asserting the info log.

Reply:
Good point about the integration test, will add it.

What would be the point of not invalidating the whole VFS?

Reply:
> What would be the point of not invalidating the whole VFS?

The penalty would be a little bit smaller since we don't invalidate things in the Gradle user home.

Reply:
The performance penalty is very severe regardless of how much of the VFS we invalidate. The sentiment here is that we don't trust the given piece of work, so it makes sense to invalidate everything. Therefore I'd leave this as is.

I've added the integration test.

Reply:
👍